core/embed/prodtest/main.c

```diff
@@ -516,6 +606,415 @@ static void test_otp_write_device_variant(const char *args) {
+      0x2d, 0xeb, 0x33, 0xa0, 0xf4, 0xa1, 0x39, 0x45, 0xd8, 0x98, 0xc2, 0x96,
+      0x4f, 0xe3, 0x42, 0xe2, 0xfe, 0x1a, 0x7f, 0x9b, 0x8e, 0xe7, 0xeb, 0x4a,
+      0x7c, 0x0f, 0x9e, 0x16, 0x2b, 0xce, 0x33, 0x57, 0x6b, 0x31, 0x5e, 0xce,
+      0xcb, 0xb6, 0x40, 0x68, 0x37, 0xbf, 0x51, 0xf5};
+  uint8_t public_key[32] = {0};
+  size_t public_key_size = 0;
+  optiga_result ret = optiga_calc_ssec(
+      OPTIGA_CURVE_P256, OID_KEY_DEV, BASE_POINT, sizeof(BASE_POINT),
+      public_key, sizeof(public_key), &public_key_size);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_printf("ERROR: optiga_calc_ssec error %d.", ret);
+    return;
+  }
+
+  vcp_printf_ex("OK: ");
+  vcp_write_as_hex(public_key, public_key_size);
+}
+
+static void keyfido_write(char *data) {
+  static const size_t EPH_PUB_KEY_SIZE = 33;
+  static const size_t PAYLOAD_SIZE = 32;
+  static const size_t CIPHERTEXT_OFFSET = EPH_PUB_KEY_SIZE;
+  static const size_t EXPECTED_SIZE = EPH_PUB_KEY_SIZE + PAYLOAD_SIZE;
+
+  // Enable key agreement usage for device key.
+
+  optiga_metadata metadata = {0};
+  uint8_t key_usage = OPTIGA_KEY_USAGE_KEYAGREE;
+  metadata.key_usage.ptr = &key_usage;
+  metadata.key_usage.len = 1;
+  metadata.execute = OPTIGA_ACCESS_ALWAYS;
+
+  if (!set_metadata(OID_KEY_DEV, &metadata)) {
+    return;
+  }
+
+  // Read encrypted FIDO attestation private key.
+
+  uint8_t data_bytes[EXPECTED_SIZE];
+  int len = get_from_hex(data_bytes, sizeof(data_bytes), data);
+  if (len < 0) {
+    vcp_printf("ERROR: Hexadecimal decoding error %d.", len);
+    return;
+  }
+
+  if (len != EXPECTED_SIZE) {
+    vcp_printf("ERROR: Unexpected input length.");
+    return;
+  }
+
+  // Expand sender's ephemeral public key.
+  curve_point pub = {0};
+  if (0 == ecdsa_read_pubkey(&nist256p1, data_bytes, &pub)) {
+    vcp_printf("ERROR: Failed to decode public key.");
+    return;
+  }
+  uint8_t public_key[4 + 64] = {0x03, 0x42, 0x00, 0x04};
+  bn_write_be(&pub.x, public_key + 4);
+  bn_write_be(&pub.y, public_key + 4 + 32);
+
+  // Execute ECDH with device private key.
+  uint8_t secret[32] = {0};
+  size_t secret_size = 0;
+  optiga_result ret = optiga_calc_ssec(OPTIGA_CURVE_P256, OID_KEY_DEV,
+                                       public_key, sizeof(public_key), secret,
+                                       sizeof(secret), &secret_size);
+  if (OPTIGA_SUCCESS != ret) {
+    memzero(secret, sizeof(secret));
+    vcp_printf("ERROR: optiga_calc_ssec error %d.", ret);
+    return;
+  }
+
+  // Hash the shared secret. Use the result as the decryption key and IV.
+  sha256_Raw(secret, secret_size, secret);
+  aes_decrypt_ctx ctx = {0};
+  AES_RETURN aes_ret = aes_decrypt_key256(secret, &ctx);
+  if (EXIT_SUCCESS != aes_ret) {
+    vcp_printf("ERROR: aes_decrypt_key256 error.");
+    memzero(&ctx, sizeof(ctx));
+    memzero(secret, sizeof(secret));
+    return;
+  }
+
+  // Decrypt the FIDO attestation key.
+  uint8_t fido_key[32] = {0};
+  uint8_t iv[16] = {0};
```

Comment:
you're not setting the IV

Reply:
The IV is intentionally all-zero. I don't see any problem in that since the encryption key is unique for each ciphertext.

Reply:
sure but then the comments should reflect that

Reply:
c77f1114a65208be5c28e667ca4addc01471c72a

Don't resolve until @onvej-sl confirms my assessment.

Reply:
what i meant is, a couple lines above a comment claims that the sha256 result is used for key _and IV_
(this comment is also good ofc)

Reply:
https://github.com/trezor/trezor-firmware/pull/3185/commits/9481fb8482a214ca14cda26974221033d14b9126

Reply:
I agree that the initialization vector can be all zeros if every key is used only once. This has to be guaranteed by the server performing the encryption.

By the way, it seems to me that it is not difficult to introduce an arbitrary initialization vector here. This would save us from these considerations.

Reply:
The server generates an ephemeral key pair for each encryption, so it's guaranteed. Let's keep it simple.

Reply:
@onvej-sl please resolve this if you're OK with the current state